x-pack/elastic-agent/pkg/core/server/server.go

```diff
@@ -151,14 +153,25 @@ func (s *Server) Start() error {
 		ClientCAs:      certPool,
 		GetCertificate: s.getCertificate,
 	})
+	s.serverLock.Lock()
 	s.server = grpc.NewServer(grpc.Creds(creds))
+	s.serverLock.Unlock()
 	proto.RegisterElasticAgentServer(s.server, s)
 
 	// start serving GRPC connections
 	go func() {
-		err := s.server.Serve(lis)
+		s.serverLock.Lock()
+		server := s.server
+		s.serverLock.Unlock()
+
+		if server == nil { // Server.Stop was called before this goroutine run
+			s.logger.Error("cannot start gRPC server, server is nil. Did you call Stop before the initialization has completed?")
+			return
+		}
+
+		err := server.Serve(lis)
```

Comment:
I'm not sure this is completely safe, `server` could still be set to `nil` here because this call is outside the mutex. `Stop()` can still be called before `Server`.

Can we do something even simpler and just not set server to `nil` on Stop()? If the agent is exiting it doesn't matter, if the server is reinitialized the old value will still be garbage collected.

Reply:
> I'm not sure this is completely safe, server could still be set to nil here because this call is outside the mutex. Stop() can still be called before Server.

@cmacknz Not sure what you mean, we save the pointer to the variable in the critical section and then check the **stored** pointer for `nil` after unlock.

The only risk we have here is to run `server.Serve` after `s.server.Stop()` which should be handled by the gRPC server implementation and return an error on `server.Serve` (my assumption, I didn't check).

Reply:
> Not sure what you mean, we save the pointer to the variable in the critical section and then check the stored pointer for nil after unlock.

Yes I looked at this too quickly between meetings. `server` can't be `nil` before calling `server.Server` now. My original point isn't valid.